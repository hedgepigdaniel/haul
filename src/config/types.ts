import webpack from 'webpack';
import { MinifyOptions } from 'terser';
import { DeepNonNullable, Overwrite, Assign } from 'utility-types';
import Runtime from '../runtime/Runtime';

export type ServerConfig = {
  port?: number;
  host?: string;
};

export type NormalizedServerConfig = DeepNonNullable<ServerConfig>;

// Options received from the CLI arguments/options.
export type EnvOptions = {
  platform: string;
  root: string;
  dev: boolean;
  bundleType?: 'basic-bundle' | 'indexed-ram-bundle' | 'file-ram-bundle';
  bundleMode: 'single-bundle' | 'multi-bundle';
  bundleTarget?: 'file' | 'server';
  assetsDest?: string;
  bundleOutput?: string;
  sourcemapOutput?: string;
  minify?: boolean;
  port?: number;
};

export type BundleConfig = Assign<
  {
    name?: string;
    entry: string | string[];
    type?: 'basic-bundle' | 'indexed-ram-bundle' | 'file-ram-bundle';
    platform?: string;
    root?: string;
    dev?: boolean;
    assetsDest?: string;
    minify?: boolean;
    minifyOptions?: Pick<
      MinifyOptions,
      Exclude<keyof MinifyOptions, 'sourceMap'>
    >;
    sourceMap?: boolean | 'inline';
    dll?: boolean;
    app?: boolean;
    dependsOn?: string[];
    providesModuleNodeModules?: Array<
      string | { name: string; directory: string }
    >;
    hasteOptions?: any;
    transform?: WebpackConfigTransform;
  },
  ExternalBundleConfig
>;

export type ExternalBundleConfig = {
  copyBundle?: boolean;
  bundlePath?: string;
  manifestPath?: string;
  assetsPath?: string;
};

export type TemplatesConfig = {
  filename: { [platform: string]: string };
};

export type NormalizedTemplatesConfig = TemplatesConfig;

export type NormalizedBundleConfig = Assign<
  Overwrite<
    Pick<
      DeepNonNullable<BundleConfig>,
      Exclude<keyof BundleConfig, 'transform' | keyof ExternalBundleConfig>
    >,
    {
      minifyOptions: BundleConfig['minifyOptions'];
    }
  >,
  {
    external:
      | false
      | Overwrite<
          DeepNonNullable<ExternalBundleConfig>,
          { manifestPath: ExternalBundleConfig['manifestPath'] }
        >;
  }
>;

export type WebpackConfigTransform = (params: {
  bundleName: string;
  config: webpack.Configuration;
  env: EnvOptions;
  runtime: Runtime;
}) => webpack.Configuration | void;

export type BundleConfigBuilder = (
  env: EnvOptions,
  runtime: Runtime
) => BundleConfig;

export type ProjectConfig = {
  server?: ServerConfig;
  platforms?: string[];
  templates?: TemplatesConfig;
  bundles: { [bundleName: string]: BundleConfigBuilder | BundleConfig };
};

export type NormalizedProjectConfig = {
  server: NormalizedServerConfig;
  platforms: string[];
  templates: NormalizedTemplatesConfig;
  bundles: { [bundleName: string]: NormalizedBundleConfig };
  webpackConfigs: { [bundleName: string]: webpack.Configuration };
};

export type NormalizedProjectConfigBuilder = (
  runtime: Runtime,
  env: EnvOptions
) => NormalizedProjectConfig;
