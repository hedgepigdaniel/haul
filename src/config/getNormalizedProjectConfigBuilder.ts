import { NormalizedProjectConfigBuilder } from './types';
import importModule from '../utils/importModule';
import Runtime from '../runtime/Runtime';

export default function getNormalizedProjectConfigBuilder(
  runtime: Runtime,
  configPath?: string
): NormalizedProjectConfigBuilder {
  let config;

  /**
   * When it doesn't have DEFAULT_CONFIG_FILENAME and it's not specified another file
   * we will use default configuration based on main file from package.json
   */
  if (!configPath) {
    throw new Error("Couldn't find `haul.config.js`");
  } else {
    config = importModule(configPath, { resolve: require.resolve, runtime })
      .exports;
    config = config.__esModule ? config.default : config;
  }

  if (typeof config !== 'function') {
    throw new Error(
      `Exported value from ${configPath} does not seem to be a valid Haul config - did you forget to use "makeConfig"?`
    );
  }

  return config as NormalizedProjectConfigBuilder;
}
