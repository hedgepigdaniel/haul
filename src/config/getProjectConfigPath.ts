import path from 'path';
import { DEFAULT_CONFIG_FILENAME } from '../constants';

export default function getProjectConfigPath(
  root: string,
  configPath?: string
) {
  return configPath
    ? path.isAbsolute(configPath)
      ? configPath
      : path.join(root, configPath)
    : path.join(root, DEFAULT_CONFIG_FILENAME);
}
