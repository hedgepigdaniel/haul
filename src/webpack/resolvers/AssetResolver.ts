import path from 'path';
import escapeStringRegexp from 'escape-string-regexp';
import Runtime from '../../runtime/Runtime';

type Request = {
  path: string;
  relativePath: string;
};

type Options = {
  test?: RegExp;
  platform: string;
  runtime: Runtime;
};

type CollectOutput = {
  [key: string]: {
    platform: string;
    name: string;
  };
};

type CollectOptions = {
  name: string;
  platform: string;
  type: string;
};

export default class AssetResolver {
  static test = /\.(aac|aiff|bmp|caf|gif|html|jpeg|jpg|m4a|m4v|mov|mp3|mp4|mpeg|mpg|obj|otf|pdf|png|psd|svg|ttf|wav|webm|webp)$/;
  static collect(
    list: Array<string>,
    { name, type, platform }: CollectOptions
  ): CollectOutput {
    const regex = /^(bmp|gif|jpg|jpeg|png|psd|tiff|webp|svg)$/.test(type)
      ? new RegExp(
          `^${escapeStringRegexp(
            name
          )}(@\\d+(\\.\\d+)?x)?(\\.(${platform}|native))?\\.${type}$`
        )
      : new RegExp(
          `^${escapeStringRegexp(name)}(\\.(${platform}|native))?\\.${type}$`
        );
    const priority = (queryPlatform: string) =>
      ['native', platform].indexOf(queryPlatform);

    // Build a map of files according to the scale
    return list.reduce(
      (acc, curr) => {
        const match = regex.exec(curr);

        if (match) {
          let [, scale, , , platform] = match;
          scale = scale || '@1x';

          if (
            acc[scale] &&
            priority(platform) < priority(acc[scale].platform)
          ) {
            // do nothing
            return acc;
          }

          return { ...acc, [scale]: { platform, name: curr } };
        }

        return acc;
      },
      {} as CollectOutput
    );
  }

  constructor(private options: Options) {}

  apply(resolver: any) {
    const platform = this.options.platform;
    const test = this.options.test || AssetResolver.test;
    const runtime = this.options.runtime;

    resolver.hooks.file.tapAsync(
      'AssetResolver',
      (request: Request, _: any, callback: Function) => {
        if (!test.test(request.path)) {
          callback();
          return;
        }

        resolver.fileSystem.readdir(
          path.dirname(request.path),
          (error: Error | null, result: any) => {
            if (error) {
              callback();
              return;
            }

            const name = path.basename(request.path).replace(/\.[^.]+$/, '');
            const type = request.path.split('.').pop() || '';

            let resolved = result.includes(path.basename(request.path))
              ? request.path
              : null;

            if (!resolved) {
              const map = AssetResolver.collect(result, {
                name,
                type,
                platform,
              });
              const key = map['@1x']
                ? '@1x'
                : Object.keys(map).sort(
                    (a, b) =>
                      Number(a.replace(/[^\d.]/g, '')) -
                      Number(b.replace(/[^\d.]/g, ''))
                  )[0];

              resolved =
                map[key] && map[key].name
                  ? path.resolve(path.dirname(request.path), map[key].name)
                  : null;
            }

            if (!resolved) {
              runtime.logger.warn(`Cannot resolve: ${request.path}`);
              callback();
              return;
            }

            const resolvedFile = {
              ...request,
              path: resolved,
              relativePath:
                request.relativePath &&
                resolver.join(request.relativePath, resolved),
              file: true,
            };

            runtime.logger.debug(
              `Resolved file: ${request.path} <--> ${resolvedFile.path}`
            );
            callback(null, resolvedFile);
          }
        );
      }
    );
  }
}
