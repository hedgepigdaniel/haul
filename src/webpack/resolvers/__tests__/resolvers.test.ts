import fs from 'fs';
import path from 'path';
import webpack from 'webpack';
import AssetResolver from '../AssetResolver';
// @ts-ignore
import config from './__fixtures__/webpack.config.js';

const filesFromEntry = [
  require.resolve('./__fixtures__/file.pdf'),
  require.resolve('./__fixtures__/file@1x.jpeg'),
  require.resolve('./__fixtures__/file@3x.png'),
  require.resolve('./__fixtures__/file@2x.gif'),
];

const runWebpack = (
  assertion: (assetPaths: Array<string>) => void,
  done: Function & { fail: (error: Error | string) => void }
) => {
  webpack(config, (err, stats) => {
    if (err) {
      done.fail(err);
    } else if (stats.hasErrors()) {
      done.fail(stats.toString());
    }
    const assetPaths = stats
      .toJson()
      .modules!.map((module: any) => module.identifier);

    try {
      assertion(assetPaths);
      done();
    } catch (error) {
      done.fail(error);
    }
  });
};

test('resolves to file@{number}x.{ext} if file.{ext} not present', done => {
  runWebpack(assetPaths => {
    expect(assetPaths).toEqual(expect.arrayContaining(filesFromEntry));
  }, done);
});

test('resolves Haste modules', done => {
  runWebpack(assetPaths => {
    expect(assetPaths).toEqual(
      expect.arrayContaining([require.resolve('./__fixtures__/HasteModule.js')])
    );
  }, done);
});

test('AssetResolver.collect returns empty object for empty list', () => {
  const result = AssetResolver.collect([], {
    name: 'filename',
    type: 'jpeg',
    platform: 'native',
  });

  expect(result).toEqual({});
});

test('AssetResolver.collect returns empty object when file not in the list', () => {
  const result = AssetResolver.collect(['file.jpeg', 'filename.png'], {
    name: 'filename',
    type: 'jpeg',
    platform: 'android',
  });

  expect(result).toEqual({});
});

test('AssetResolver.collect returns files based on requested platform', () => {
  const fileList = [
    'mail.android.png',
    'mail.png',
    'mail@2x.android.png',
    'mail@2x.png',
    'mail@3x.android.png',
    'mail@3x.png',
    'mail.ios.png',
    'mail.native.png',
  ];
  const fileListDifferentOrder = [
    'mail.ios.png',
    'mail.native.png',
    'mail.png',
    'mail.android.png',
    'mail@2x.png',
    'mail@2x.android.png',
    'mail@3x.png',
    'mail@3x.android.png',
  ];

  [fileList, fileListDifferentOrder].forEach(files => {
    const result = AssetResolver.collect(files, {
      name: 'mail',
      type: 'png',
      platform: 'android',
    });
    expect(result).toEqual({
      '@1x': {
        name: 'mail.android.png',
        platform: 'android',
      },
      '@2x': {
        name: 'mail@2x.android.png',
        platform: 'android',
      },
      '@3x': {
        name: 'mail@3x.android.png',
        platform: 'android',
      },
    });
  });
});

test('AssetResolver.collect returns a map of paths to resolve', () => {
  const files = fs.readdirSync(path.resolve(__dirname, './__fixtures__'));
  const result = AssetResolver.collect(files, {
    name: 'filename',
    type: 'jpeg',
    platform: 'ios',
  });

  expect(result).toMatchSnapshot();
});
