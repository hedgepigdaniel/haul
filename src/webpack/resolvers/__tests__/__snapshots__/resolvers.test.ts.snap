// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssetResolver.collect returns a map of paths to resolve 1`] = `
Object {
  "@1x": Object {
    "name": "filename@1x.ios.jpeg",
    "platform": "ios",
  },
  "@2x": Object {
    "name": "filename@2x.jpeg",
    "platform": undefined,
  },
  "@3x": Object {
    "name": "filename@3x.jpeg",
    "platform": undefined,
  },
  "@4x": Object {
    "name": "filename@4x.native.jpeg",
    "platform": "native",
  },
}
`;
